File: Kata 6/JuegoDelPescador/JuegoDelPescador.cs
/* Kata Level: 6.
 * 
 * Write a simple parser that will parse and run Deadfish.
 * 
 * Deadfish has 4 commands, each 1 character long:
 * 
 * i increments the value (initially 0)
 * d decrements the value
 * s squares the value
 * o outputs the value into the return array
 * Invalid characters should be ignored.
 * 
 * Deadfish.parse("iiisdoso") =- new int[] {8, 64};
 */

using System.Collections.Generic;
using System.Linq;

public class Deadfish
{
    public static int[] Parse(string data)
    {
        int sum = 0;
        List<int> result = new List<int>();
        int[] array = new int[data.Where(x => x == 'o').Count()];

        foreach(char letter in data)
        {
            if(letter == 'i')   		sum += 1;
            else if(letter == 'd')    	sum -= 1;
            else if(letter == 's')    	sum *= sum;
            else if(letter == 'o')		result.Add(sum);
        }
        for(int i = 0 ; i < result.Count ; i++)    array[i] = result[i];
        return array;               // Alternativa:     Evitamos crear el array a la vez que la lista, y en la última línea escribimos      return result.ToArray();
    }

    public static void Main(string[] args)
    {
        foreach(int number in Parse("iiisdoso"))	Console.WriteLine(number);
    }
}

File: Kata 6/JuegoDelPescador/JuegoDelPescador.java
/* Kata Level: 6.
 * 
 * Write a simple parser that will parse and run Deadfish.
 * 
 * Deadfish has 4 commands, each 1 character long:
 * 
 * i increments the value (initially 0)
 * d decrements the value
 * s squares the value
 * o outputs the value into the return array
 * Invalid characters should be ignored.
 * 
 * Deadfish.parse("iiisdoso") =- new int[] {8, 64};
 */

import java.util.*;

public class JuegoDelPescador {

	public static int[] parse(String data) {
        List<Integer> lista = new ArrayList<>();
        int contador = 0;
        for(int i = 0 ; i < data.length() ; i++) {
            if(data.charAt(i) == 'i')
                contador++;
            else if(data.charAt(i) == 'd')
                contador--;
            else if(data.charAt(i) == 's')
                contador *= contador;
            else if(data.charAt(i) == 'o')
                lista.add(contador);
        }
        int[] solucion = new int[lista.size()];
        for(int i = 0 ; i < lista.size() ; i++) 
            solucion[i] = lista.get(i);
        return solucion;
	}
	 
	public static void main(String[] args) {
		
		int[] numeros = parse("iiisdoso");
		for(int numero : numeros)
			System.out.print(numero + " ");

	}

}